File: HW2.js
// jshint esversion:6
const prompt = require("prompt-sync")();
let memoryAddressSpace = prompt(
	"Enter the memory address type: (1:bits, 2:bytes, 3:kilobytes, 4:megabytes, 5:gigabytes): ",
);
switch (memoryAddressSpace) {
	case "1":
		memoryAddressSpace = prompt("Enter the memory address space: ");
		break;
	case "2":
		memoryAddressSpace = prompt("Enter the memory address space: ");
		memoryAddressSpace *= 8;
		break;
	case "3":
		memoryAddressSpace = prompt("Enter the memory address space: ");
		memoryAddressSpace = memoryAddressSpace * 8 * 1024;
		break;
	case "4":
		memoryAddressSpace = prompt("Enter the memory address space: ");
		memoryAddressSpace = memoryAddressSpace * 8 * 1024 * 1024;
		break;
	case "5":
		memoryAddressSpace = prompt("Enter the memory address space: ");
		memoryAddressSpace = memoryAddressSpace * 8 * 1024 * 1024 * 1024;
		break;
	default:
		console.log("Invalid input");
}
const cacheSize = prompt("Enter the cache size: ");
const blockSize = prompt("Enter the block size: ");
// memoryAddressSpace = Math.round(Math.log2(memoryAddressSpace));
const numberOfBlocks = (blockSize) => {
	return Math.round(Math.log2(blockSize));
};
const numberOfSets = (cacheSize, blockSize) => {
	return Math.round(Math.log2(cacheSize / blockSize));
};
const numberOfTags = (memoryAddressSpace, set, blockSize) => {
	return memoryAddressSpace - set - blockSize;
};
console.log(`\n Direct Mapping Method: \n`);
let Arr = [];
Arr.Block = numberOfBlocks(blockSize);
Arr.Set = numberOfSets(cacheSize * 1024, blockSize);
Arr.Tag = numberOfTags(memoryAddressSpace, Arr.Set, Arr.Block);
console.table(Arr);
console.log(`\n 2-way associative Mapping Method: \n`);
Arr = [];
Arr.Block = numberOfBlocks(blockSize);
Arr.Set = numberOfSets(cacheSize * 1024, blockSize);
Arr.Tag = numberOfTags(memoryAddressSpace, Arr.Set, Arr.Block);
Arr.Set -= 1;
Arr.Tag += 1;
console.table(Arr);
console.log(`\n 4-way associative Mapping Method: \n`);
Arr = [];
Arr.Block = numberOfBlocks(blockSize);
Arr.Set = numberOfSets(cacheSize * 1024, blockSize);
Arr.Tag = numberOfTags(memoryAddressSpace, Arr.Set, Arr.Block);
Arr.Set -= 2;
Arr.Tag += 2;
console.table(Arr);
console.log(`\n 8-way associative Mapping Method: \n`);
Arr = [];
Arr.Block = numberOfBlocks(blockSize);
Arr.Set = numberOfSets(cacheSize * 1024, blockSize);
Arr.Tag = numberOfTags(memoryAddressSpace, Arr.Set, Arr.Block);
Arr.Set -= 3;
Arr.Tag += 3;
console.table(Arr);
console.log(`\n Fully associative Mapping Method: \n`);
Arr = [];
Arr.Block = numberOfBlocks(blockSize);
Arr.Tag = numberOfTags(
	memoryAddressSpace,
	numberOfSets(cacheSize * 1024, blockSize),
	Arr.Block,
);
const temp = numberOfSets(cacheSize * 1024, blockSize);
Arr.Tag += temp;
console.table(Arr);

File: HW2.cpp
#include <iostream>
#include <cmath>
using namespace std;
int numberOfBlocks(int blockSize)
{
   return round(log2(blockSize));
}
int numberOfSets(int cacheSize, int blockSize)
{
   return round(log2(cacheSize / blockSize));
}
int numberOfTags(int memoryAddressSpace, int set, int blockSize)
{
   return memoryAddressSpace - set - numberOfBlocks(blockSize);
}
int main()
{
   cout << "Enter the memory address space: ";
   int memoryAddressSpace = 32;
   cin >> memoryAddressSpace;
   cout << "Enter the cacheSize size: ";
   int cacheSize = 256;
   cin >> cacheSize;
   cout << "Enter the block size: ";
   int blockSize = 128;
   cin >> blockSize;
   int block = numberOfBlocks(blockSize);
   int set = numberOfSets(cacheSize * 1024, blockSize);
   int tag = numberOfTags(memoryAddressSpace, set, blockSize);
   cout << "Direct Mapping :" << endl;
   cout << "Number of blocks: " << block << endl;
   cout << "Number of sets: " << set << endl;
   cout << "Number of tags: " << tag << endl;
   cout << "2-way Associative Mapping :" << endl;
   set = 0;
   tag = 0;
   set = numberOfSets(cacheSize * 1024, blockSize);
   tag = numberOfTags(memoryAddressSpace, set, blockSize);
   set = set - 1;
   tag = tag + 1;
   cout << "Number of blocks: " << numberOfBlocks(blockSize) << endl;
   cout << "Number of sets: " << set << endl;
   cout << "Number of tags: " << tag << endl;
   cout << "4-way Associative Mapping :" << endl;
   set = 0;
   tag = 0;
   set = numberOfSets(cacheSize * 1024, blockSize);
   tag = numberOfTags(memoryAddressSpace, set, blockSize);
   set = set - 2;
   tag = tag + 2;
   cout << "Number of blocks: " << numberOfBlocks(blockSize) << endl;
   cout << "Number of sets: " << set << endl;
   cout << "Number of tags: " << tag << endl;
   cout << "8-way Associative Mapping :" << endl;
   set = 0;
   tag = 0;
   set = numberOfSets(cacheSize * 1024, blockSize);
   tag = numberOfTags(memoryAddressSpace, set, blockSize);
   set = set - 3;
   tag = tag + 3;
   cout << "Number of blocks: " << numberOfBlocks(blockSize) << endl;
   cout << "Number of sets: " << set << endl;
   cout << "Number of tags: " << tag << endl;
   cout << "Fully Associative Mapping :" << endl;
   set = 0;
   tag = 0;
   set = numberOfSets(cacheSize * 1024, blockSize);
   tag = numberOfTags(memoryAddressSpace, set, blockSize);
   tag = tag + set;
   cout << "Number of blocks: " << numberOfBlocks(blockSize) << endl;
   cout << "Number of tags: " << tag << endl;
   return 0;
}
